### ISSUE TITLE
Yolov11 custom python oonx vs ultralytics CLI produce different results

### ISSUE THREAD

**Issue by 913-Gaina-Florin** [ultralytics/ultralytics#20521]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

import onnxruntime as ort
import numpy as np
import cv2
 
 
 
class YoloInference:
    # List of class names for YOLO model
    classes = [
    'speed20', 'speed30', 'speed50', 'speed60', 'speed70', 'speed80', 'speed100', 'speed120'
    ]
    
    def __init__(self, model_path):
        self.session = ort.InferenceSession(model_path)
        self.input_name = self.session.get_inputs()[0].name
        self.output_name = self.session.get_outputs()[0].name
 
    def preprocess(self, image):
        """This functions preprocess the image before inference
        preprocesses:resize(640x640),normalize,HWC to CHW,adds batch
        Returns:
            img preprocessed
        """
        img = cv2.resize(image, (640, 640))  # Resize to model input size
        img = img.astype(np.float32)
        img = img / 255.0  # Normalize
        img = np.transpose(img, [2, 0, 1])  # Change data layout from HWC to CHW
        img = np.expand_dims(img, axis=0)  # Add batch dimension
        return img
 
    def postprocess(self, output, conf_threshold=0.5, nms_threshold=0.1):
        detections = output[0].T  # Transpose the array
 
        boxes = []
        confidences = []
        class_ids = []
 
        for detection in detections:  # Iterate through each of the 8400 predictions
            box = detection[:4]
            class_scores = detection[4:]
 
            class_id = np.argmax(class_scores)
            confidence = class_scores[class_id]
            if confidence > conf_threshold:
                # Convert from center coordinates to corner coordinates
                cx, cy,box_width, box_height = box  # Note the order: x, y, height, width
                cx = cx.item()
                cy = cy.item()
                box_height = box_height.item()
                box_width = box_width.item()
                x1 = int((cx - box_width / 2))
                y1 = int((cy - box_height / 2))
                x2 = int(cx + box_width / 2)
                y2 = int(cy + box_height / 2)
                boxes.append([x1, y1, x2, y2])
                confidences.append(float(confidence.item()))
                class_ids.append(class_id)
 
        # Apply non-maximum suppression (NMS)
        indices = cv2.dnn.NMSBoxes(boxes, confidences, conf_threshold, nms_threshold)
        
        print(confidences)
        
        processed_results = []
        if len(indices) > 0:
            for i in indices.flatten():
                x1, y1, x2, y2 = boxes[i]
                processed_results.append({
                    'class_id': self.classes[class_ids[i]],
                    'confidence': confidences[i],
                    'box': [x1, y1, x2, y2]
                })
        return processed_results
 
    def infer(self, image,desired_class_id=None): # 77-tedy 21-bear
        h,w,_ = image.shape
        input_image = self.preprocess(image)
        output = self.session.run([self.output_name], {self.input_name: input_image})
        processed_results = self.postprocess(output)
        orig_height, orig_width = image.shape[:2]
        scale_x = orig_width / w
        scale_y = orig_height / h
        if desired_class_id is not None:
            boxes = [result['box'] for result in processed_results if result['class_id']==self.classes[desired_class_id]]
            scores = [result['confidence'] for result in processed_results if result['class_id']==self.classes[desired_class_id]]
            class_ids = [result['class_id'] for result in processed_results if result['class_id']==self.classes[desired_class_id]]
        else:
            boxes = [result['box'] for result in processed_results]
            scores = [result['confidence'] for result in processed_results]
            class_ids = [result['class_id'] for result in processed_results]
        return boxes, scores, class_ids, scale_x, scale_y
 
    def draw_detections(self,image, boxes, scores, class_ids, scale_x, scale_y):
        for box, score, class_id in zip(boxes, scores, class_ids):
            x1, y1, x2, y2 = box
 
            # Scale the coordinates back to the original image size
            x1 = int(x1 * scale_x)
            y1 = int(y1 * scale_y)
            x2 = int(x2 * scale_x)
            y2 = int(y2 * scale_y)
            
            print(x1, y1, x2, y2)
 
            # Ensure class_id is used correctly to fetch class name
            class_name = class_id if isinstance(class_id, str) else self.classes[class_id]
 
            # Draw the bounding box
            cv2.rectangle(image, (x1, y1), (x2, y2), (0, 255, 0), 2)
 
            # Label the bounding box with class name and confidence score
            label = f'{class_name}: {score:.2f}'
            cv2.putText(image, label, (x1, y1 - 10), cv2.FONT_HERSHEY_SIMPLEX, 0.9, (36, 255, 12), 2)
 
        return image
 
    # Load and run inference
    yolo = YoloInference(model_path)
    image = cv2.imread(image_path)
    boxes, scores, class_ids, scale_x, scale_y = yolo.infer(image)
 
    # Draw and display detections
    image_with_detections = yolo.draw_detections(image, boxes, scores, class_ids, scale_x, scale_y)
    cv2.imwrite("Detectionsl.jpg", image_with_detections)

Hello! I am trying to run an exported verision of a fine-tunned yolov11n to detect roadisgns. When i run the model with the yolo cli on this exact same image, I get a object detected with confidence 0.86, but when i run it with my custom script, I get no meaningful detections, all confidence values are around 1e-7. Can anyoane please help me, I have been struggling to solve this for the past few days, I tried tensorflow, Pillow, opencv, differnet normaliations for preprocessing, but nothing seems to work.

![Image](https://github.com/user-attachments/assets/36d8e960-c34e-4d7f-8bda-9888b3f23aac)

![Image](https://github.com/user-attachments/assets/f469ee26-474c-42db-8158-47f15dcf5581)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @913-Gaina-Florin, thank you for reaching out and sharing your detailed code and observations with the Ultralytics team! 🚦🔎 This is an automated response to help you get the most effective support, and an Ultralytics engineer will assist you soon.

We recommend reviewing the [Docs](https://docs.ultralytics.com/) for guidance on ONNX model export and inference, as well as checking the available [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples. Many common questions and troubleshooting steps are covered there.

Since this appears to be a 🐛 Bug Report (inference result mismatch), could you please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) if the code above does not already qualify? This helps us debug and compare results more effectively.

A few suggestions as you troubleshoot:
- Double-check that your ONNX export was performed with the latest `ultralytics` package.
- Ensure the input preprocessing in your custom script matches the preprocessing expected by the YOLO11 ONNX model. Even small differences (such as channel order, normalization, or image resizing) can cause large discrepancies.
- Compare your postprocessing (especially confidence thresholding and NMS) to the Ultralytics reference implementation.

Join the Ultralytics community for additional support:
- Real-time chat: [Discord](https://discord.com/invite/ultralytics) 🎧
- In-depth discussions: [Discourse](https://community.ultralytics.com/)
- Share knowledge: [Subreddit](https://reddit.com/r/Ultralytics)

## Upgrade

Please upgrade to the latest `ultralytics` package (and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml)) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to ensure your issue is not already resolved:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of these verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you again for your report and patience! An Ultralytics engineer will review your issue and follow up with you soon.

**Comment by Y-T-G** [comment]
The preprocessing is wrong.

You can check the ONNX example here:
https://github.com/ultralytics/ultralytics/tree/main/examples
